From 5e10edc1cdc88caabccfb6076e02e3c0ad5d2402 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Tue, 17 Dec 2024 23:25:29 +0100
Subject: [PATCH 85/86] ntdll: Avoid merging of memcpy and memmove.

They are identical in Wine, maybe therefore the compiler merges
them into a single _memcpy, so both can handle overlapping areas.

But ASan does not know Wine's memcpy can handle it
and therefore triggers the error.


=================================================================
==640==ERROR: AddressSanitizer: memcpy-param-overlap: memory ranges [0x00857b20,0x00857b2e) and [0x00857b21, 0x00857b2f) overlap
02e4:fixme:msvcrt:_set_abort_behavior _WRITE_CALL_REPORTFAULT unhandled
02e4:fixme:file:server_get_file_info Unsupported info class e
    #0 0x76e323d1 in memcpy /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/../sanitizer_common/sanitizer_common_interceptors_memintrinsics.inc:115:5
    #1 0x7891968e in RtlMoveMemory /usr/src/packages/BUILD\dlls/ntdll\string.c:265:13
    #2 0x00655e6e in test_RtlMoveMemory /home/bernhard/data/entwicklung/2024/wine\wine/dlls/ntdll/tests/rtl.c:302:22
    #3 0x00655e6e in func_rtl /home/bernhard/data/entwicklung/2024/wine\wine/dlls/ntdll/tests/rtl.c:4077:5
    #4 0x00776711 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #5 0x00776711 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #6 0x0077868b in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #7 0x7877ea3f in BaseThreadInitThunk (C:\windows\system32\kernel32.dll+0x7b80ea3f)
    #8 0x788fce82 in call_thread_func_wrapper (C:\windows\system32\ntdll.dll+0x7bc0ce82)
    #9 0x789304b4 in call_thread_func /usr/src/packages/BUILD\dlls/ntdll\signal_i386.c:524:9

Address 0x00857b20 is a wild pointer inside of access range of size 0x0000000e.
Address 0x00857b21 is a wild pointer inside of access range of size 0x0000000e.
SUMMARY: AddressSanitizer: memcpy-param-overlap /usr/src/packages/BUILD\dlls/ntdll\string.c:265:13 in RtlMoveMemory
==640==ABORTING
make: *** [Makefile:305163: dlls/ntdll/tests/i386-windows/rtl.ok] Fehler 1



$ i686-w64-mingw32-objdump --disassemble build-32/obj/dlls/ntdll/i386-windows/ntdll.dll | grep -E "RtlMoveMemory|RtlCopyMemory" -A13
7bc19500 <_RtlCopyMemory@12>:
7bc19500:       55                      push   %ebp
7bc19501:       89 e5                   mov    %esp,%ebp
7bc19503:       83 ec 0c                sub    $0xc,%esp
7bc19506:       8b 45 10                mov    0x10(%ebp),%eax
7bc19509:       89 44 24 08             mov    %eax,0x8(%esp)
7bc1950d:       8b 45 0c                mov    0xc(%ebp),%eax
7bc19510:       89 44 24 04             mov    %eax,0x4(%esp)
7bc19514:       8b 45 08                mov    0x8(%ebp),%eax
7bc19517:       89 04 24                mov    %eax,(%esp)
7bc1951a:       e8 11 71 03 00          call   7bc50630 <_memcpy>
7bc1951f:       c9                      leave
7bc19520:       c2 0c 00                ret    $0xc
7bc19523:       90                      nop
--
7bc29670 <_RtlMoveMemory@12>:
7bc29670:       55                      push   %ebp
7bc29671:       89 e5                   mov    %esp,%ebp
7bc29673:       83 ec 0c                sub    $0xc,%esp
7bc29676:       8b 45 10                mov    0x10(%ebp),%eax
7bc29679:       89 44 24 08             mov    %eax,0x8(%esp)
7bc2967d:       8b 45 0c                mov    0xc(%ebp),%eax
7bc29680:       89 44 24 04             mov    %eax,0x4(%esp)
7bc29684:       8b 45 08                mov    0x8(%ebp),%eax
7bc29687:       89 04 24                mov    %eax,(%esp)
7bc2968a:       e8 a1 6f 02 00          call   7bc50630 <_memcpy>
7bc2968f:       c9                      leave
7bc29690:       c2 0c 00                ret    $0xc
7bc29693:       90                      nop



$ i686-w64-mingw32-gcc --version
i686-w64-mingw32-gcc (GCC) 12-win32
---
 dlls/ntdll/string.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/string.c b/dlls/ntdll/string.c
index 46fb80a3a69..5881c047c32 100644
--- a/dlls/ntdll/string.c
+++ b/dlls/ntdll/string.c
@@ -125,8 +125,9 @@ void * __cdecl memcpy( void *dst, const void *src, size_t n )
  */
 void * __cdecl memmove( void *dst, const void *src, size_t n )
 {
-    volatile unsigned char *d = dst;  /* avoid gcc optimizations */
     const unsigned char *s = src;
+    volatile unsigned char *d = dst;  /* avoid gcc optimizations */
+                                      /* changed order to avoid merge with memcpy */
 
     if ((size_t)dst - (size_t)src >= n)
     {
-- 
2.47.1

